ega-data-api-htsget/src/main/java/eu/elixir/ega/ebi/htsget/formats/AbstractDataProvider.java

```diff
@@ -0,0 +1,141 @@
+        Optional<Integer> startBlockEndTrim = Optional.empty();
+        if (endBlockPosition == startBlockPosition) {
+            startBlockEndTrim = Optional.of(BlockCompressedFilePointerUtil.getBlockOffset(chunkEnd));
+        }
+
+        results.add(new HtsgetUrlV2(getTrimmedBlockAsDataUri(startBlockPosition,
+                startBlockLength,
+                startBlockStartTrim,
+                startBlockEndTrim,
+                dataStream)));
+
+        if (startBlockPosition == endBlockPosition)
+            return results;
+
+        // blocks in the middle will never have a problem with records that are at the edge of the block
+        // because the other part of the record is in a block that is included too. so in the middle it
+        // is OK to return one big range of bytes.
+        long midBlockPosition = startBlockPosition + startBlockLength;
+        if (midBlockPosition < endBlockPosition) {
+            HtsgetUrlV2 chunkURL = new HtsgetUrlV2(baseURI, "body");
+            HttpRange range = HttpRange.createByteRange(midBlockPosition, endBlockPosition - 1);
+            chunkURL.setHeader("Range", "bytes=" + range.toString());
+            results.add(chunkURL);
+        }
+
+        // Make the end block
+        int endBlockLength = AbstractDataProvider.getBlockSize(dataStream, endBlockPosition);
+        Optional<Integer> endBlockEndTrim = Optional.of(BlockCompressedFilePointerUtil.getBlockOffset(chunkEnd) + 1);
+
+       results.add(new HtsgetUrlV2(getTrimmedBlockAsDataUri(endBlockPosition,
+                endBlockLength,
+                Optional.empty(),
+                endBlockEndTrim,
+                dataStream)));
+
+       return results;
+    }
+
+    private URI getTrimmedBlockAsDataUri(long startBlockPosition, int blockSize, Optional<Integer> startTrim, Optional<Integer> endTrim, SeekableStream dataStream) throws IOException, URISyntaxException {
+
+        // Read the compressed block
+        byte[] originalBlock = new byte[blockSize];
+        dataStream.seek(startBlockPosition);
+        if (dataStream.read(originalBlock) != blockSize)
+            throw new IOException("Failed to read block");
+
+        if ((((int) originalBlock[0]) & 0xff) != BlockCompressedStreamConstants.GZIP_ID1 || (((int) originalBlock[1]) & 0xff) != BlockCompressedStreamConstants.GZIP_ID2)
+            throw new RuntimeException("Invalid block (expected header bytes " + BlockCompressedStreamConstants.GZIP_ID1 + "," + BlockCompressedStreamConstants.GZIP_ID2 + " but got " + originalBlock[0] + "," + originalBlock[1]);
+
+        // Decompress it
+        byte[] uncompressedData = blockDecompress(originalBlock);
+
+        // Trim it
+        if (endTrim.isPresent())
+            uncompressedData = Arrays.copyOf(uncompressedData, endTrim.get());
+        if (startTrim.isPresent())
+            uncompressedData = Arrays.copyOfRange(uncompressedData, startTrim.get(), uncompressedData.length);
```

Comment:
Just an idea if you want to simplify it and make only one array copy. I may be off sum a 1 or so. Check the math if you are gonna put it. This is just a thought I had while rereading the code.

```suggestion
uncompressedData = trim(uncompressedData, startTrim.orElse(0), startTrim.orElse(uncompressedData.length))

private byte[] trim(byte[] data, int start , int end){
     return Arrays.copyOfRange(data,start, end-start);
```